File: rumbo-api/app/shared_kernel/geometry.py
"""Minimal planar geometry shared by bounded contexts (no framework dependencies)."""
from __future__ import annotations


def _cross(ax: float, ay: float, bx: float, by: float, cx: float, cy: float) -> float:
    return (bx - ax) * (cy - ay) - (by - ay) * (cx - ax)


def segments_properly_cross(
    ax: float, ay: float, bx: float, by: float, cx: float, cy: float, dx: float, dy: float
) -> bool:
    """True only when segment AB passes from one side of CD to the other.

    Endpoint touches and collinear overlap are intentionally ignored so a route can end
    at a doorway or follow a boundary without counting as a through-wall traversal.
    """
    c1 = _cross(ax, ay, bx, by, cx, cy)
    c2 = _cross(ax, ay, bx, by, dx, dy)
    c3 = _cross(cx, cy, dx, dy, ax, ay)
    c4 = _cross(cx, cy, dx, dy, bx, by)
    epsilon = 1e-7
    return (c1 * c2 < -epsilon) and (c3 * c4 < -epsilon)

File: rumbo-web-app/src/modules/modeling/domain/geometry.ts
import type { Point2D, WallSegment } from './model'

const cross = (a: Point2D, b: Point2D, c: Point2D) =>
  (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x)

/**
 * True only when two segments properly cross each other. Endpoint touches and
 * collinear overlap are intentionally allowed so a route can terminate at a doorway
 * or follow a boundary without being treated as a through-wall traversal.
 */
export function segmentsProperlyCross(a: Point2D, b: Point2D, c: Point2D, d: Point2D): boolean {
  const c1 = cross(a, b, c)
  const c2 = cross(a, b, d)
  const c3 = cross(c, d, a)
  const c4 = cross(c, d, b)
  const epsilon = 1e-7
  return c1 * c2 < -epsilon && c3 * c4 < -epsilon
}

export function segmentCrossesWalls(a: Point2D, b: Point2D, walls: WallSegment[]): boolean {
  return walls.some((wall) => segmentsProperlyCross(a, b, wall.start, wall.end))
}

export function clampPoint(point: Point2D, width: number, height: number): Point2D {
  return {
    x: Math.max(0, Math.min(width, point.x)),
    y: Math.max(0, Math.min(height, point.y)),
  }
}
